### PR TITLE
[cli] fix signal handling

### PR DESCRIPTION
**Pull request by wgtdkp**
Synchronization primitives (std::mutex, std::future) in namespace `std` are used to synchronize calls to the commissioner API. Unfortunately, those synchronization primitives are not [signal-safe](https://man7.org/linux/man-pages/man7/signal-safety.7.html) and will result in deadlock.

This PR fixes this by using a dedicated thread to wait for the signal.

### CODE REVIEW

File: src/app/cli/main.cpp [hunk 1]
```diff
@@ -76,30 +78,41 @@ static void PrintVersion()
...
-static void HandleSignalInterrupt(int)
+static void HandleSignalInterrupt()
 {
-    gInterpreter.AbortCommand();
+    int signalNum = 0;
+
+    while (true)
+    {
+        sigwait(&gSignalSet, &signalNum);
+        gInterpreter.AbortCommand();
```
Comment (simonlingoogle) [inline src/app/cli/main.cpp:90]: should we quit this thread after this?

Comment (wgtdkp) [reply src/app/cli/main.cpp:90]: This Thread is designed to run in background to handle the interrupt signals. As the user can input `CRTL + C` in the CLI for multiple times, we need to keep this thread running to handle more interrupts (The name `AbortCommand` is misleading, it should be `CancelCommand` and I have fired a PR https://github.com/openthread/ot-commissioner/pull/135 to rename it).

Comment (simonlingoogle) [reply src/app/cli/main.cpp:90]: nit: If this thread goes on, you need to make sure that `AbortCommand` is reentrant.

Comment (wgtdkp) [reply src/app/cli/main.cpp:90]: Sure, it is.

**Comment by jwhui** [comment]
@wgtdkp , please help resolve conflicts.

**Comment by wgtdkp** [comment]
@jwhui done.

File: src/app/cli/main.cpp [hunk 2]
```diff
@@ -76,30 +78,41 @@ static void PrintVersion()
...
     {
         PrintVersion();
-        return 0;
+        ExitNow();
     }
 
-    signal(SIGINT, HandleSignalInterrupt);
+    // Block signals in this thread and subsequently spawned threads.
+    sigemptyset(&gSignalSet);
+    sigaddset(&gSignalSet, SIGINT);
```
Comment (bukepo) [inline src/app/cli/main.cpp:112]: is SIGINT the only one should take care of?

**Pull request merged**